### PR TITLE
[4.0 dev] Replace minicolors with color-picker

### PR DESCRIPTION
**Pull request by thednp**
Pull Request for Issue #32494.

### Summary of Changes
* removed most `minicolors` assets
* added `tinycolor` asset
* added the new `joomla-field-color-picker.es6.js` as described in #32494
* updated the `layouts/joomla/form/field/color/advanced.php` file
* updated 3 language files
* updated most assets related json files with the new assets

### Testing Instructions
This is an initial commit, as is I don't expect the compilers to do magic right now. @dgrammatiko will probably do that.
Once the compilers work, I will update this section.

### Actual result BEFORE applying this Pull Request
All `type="color"` fields that use `control="advanced"` are outdated in terms of accessibility, usability and because they require jQuery.

### Expected result AFTER applying this Pull Request
All your `type="color"` fields that use `control="advanced"` are super awesome on any device and any user.

### Documentation Changes Required
Yes. We must explain to developers how it works so they don't get funny ideas.

### CODE REVIEW

File: build/media_source/system/joomla.asset.json [hunk 1]
```diff
@@ -92,14 +92,33 @@
...
+      "dependencies": [
+        "tinycolor#script"
+      ],
+      "package": "@ctrl/tinycolor",
+      "version": "3.4.0"
+    },    
+    {
+      "name": "field.color-picker",
+      "type": "script",
+      "uri": "system/fields/joomla-field-color-picker.es6.min.js",
```
Comment (dgrammatiko) [inline build/media_source/system/joomla.asset.json]: change this to `"system/fields/joomla-field-color-picker.min.js"` we're trying to standirdise the es6 to be the clean filename

Comment (thednp) [reply build/media_source/system/joomla.asset.json]: And now?

File: administrator/language/en-GB/joomla.ini [hunk 2]
```diff
@@ -226,6 +226,16 @@ JFIELD_COLOR_LABEL_SLIDER_SATURATION="Saturation Slider"
 JFIELD_COLOR_SELECT="Select a colour"
 JFIELD_COLOR_TRANSPARENT="No colour, transparent"
 JFIELD_COLOR_VALUE="Colour with hexadecimal value of"
+JFIELD_COLOR_ADVANCED_FORMAT_HEX="HEX color format"
```
Comment (brianteeman) [inline administrator/language/en-GB/joomla.ini]:
```suggestion
JFIELD_COLOR_ADVANCED_FORMAT_HEX="HEX colour format"
```

Comment (thednp) [reply administrator/language/en-GB/joomla.ini]: ah yes, `en-GB`

File: administrator/language/en-GB/joomla.ini [hunk 3]
```diff
@@ -226,6 +226,16 @@ JFIELD_COLOR_LABEL_SLIDER_SATURATION="Saturation Slider"
...
+JFIELD_COLOR_ADVANCED_FORMAT_RGB="RGB color format"
```
Comment (brianteeman) [inline administrator/language/en-GB/joomla.ini]:
```suggestion
JFIELD_COLOR_ADVANCED_FORMAT_RGB="RGB colour format"
```

File: administrator/language/en-GB/joomla.ini [hunk 4]
```diff
@@ -226,6 +226,16 @@ JFIELD_COLOR_LABEL_SLIDER_SATURATION="Saturation Slider"
...
+JFIELD_COLOR_ADVANCED_FORMAT_HSL="HSL color format"
```
Comment (brianteeman) [inline administrator/language/en-GB/joomla.ini]:
```suggestion
JFIELD_COLOR_ADVANCED_FORMAT_HSL="HSL colour format"
```

File: api/language/en-GB/joomla.ini [hunk 5]
```diff
@@ -223,6 +223,16 @@ JFIELD_COLOR_LABEL_SLIDER_SATURATION="Saturation Slider"
 JFIELD_COLOR_SELECT="Select a colour"
 JFIELD_COLOR_TRANSPARENT="No colour, transparent"
 JFIELD_COLOR_VALUE="Colour with hexadecimal value of"
+JFIELD_COLOR_ADVANCED_FORMAT_HEX="HEX color format"
```
Comment (brianteeman) [inline api/language/en-GB/joomla.ini]: Same changes as above please

File: language/en-GB/joomla.ini [hunk 6]
```diff
@@ -179,6 +179,16 @@ JFIELD_CATEGORY_DESC="Category"
 JFIELD_COLOR_SELECT="Select a colour"
 JFIELD_COLOR_TRANSPARENT="No colour, transparent"
 JFIELD_COLOR_VALUE="Colour with hexadecimal value of"
+JFIELD_COLOR_ADVANCED_FORMAT_HEX="HEX color format"
```
Comment (brianteeman) [inline language/en-GB/joomla.ini]: same changes as above please

File: layouts/joomla/form/field/color/advanced.php [hunk 7]
```diff
@@ -3,13 +3,14 @@
  * @package     Joomla.Site
  * @subpackage  Layout
  *
- * @copyright   Copyright (C) 2005 - 2020 Open Source Matters, Inc. All rights reserved.
+ * @copyright   (C) 2005 - 2021 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (brianteeman) [inline layouts/joomla/form/field/color/advanced.php]: Do not change this line please

Comment (thednp) [reply layouts/joomla/form/field/color/advanced.php]: You already changed that back, I don't need to do anything about this right?

**Review by thednp — COMMENTED** [review]
Looks good to me.

**Comment by dgrammatiko** [comment]
@thednp you need to rebase your PR, there are conflicts (meaning the code in this PR has a base that is outdated). @richard67 could you help here?

**Comment by thednp** [comment]
It says something about write access.

File: layouts/joomla/form/field/color/advanced.php [hunk 8]
```diff
@@ -42,56 +43,133 @@
...
-if ($validate !== 'color' && in_array($format, array('rgb', 'rgba'), true))
+// set hsla / rgba to short format
+// both tinycolor and <color-picker> handle alpha automatically
+// an invalid format defaults to 'hex'
+$format = in_array($format, array('hex', 'rgb', 'hsl', 'rgba', 'hsla'), true) ? str_replace('a', '', $format) : 'hex';
+
+// set a generic placeholder
+if ($validate !== 'color' && $format === 'rgb')
+{
+  $placeholder = 'rgba(0, 0, 0, 0.5) / rgb(0, 0, 0)';
```
Comment (Quy) [inline layouts/joomla/form/field/color/advanced.php]: Please use tabs and not spaces to indent.

**Comment by richard67** [comment]
> @thednp you need to rebase your PR, there are conflicts (meaning the code in this PR has a base that is outdated). @richard67 could you help here?

@dgrammatiko I can try later if still necessary then i.e. if @thednp can't solve it himself.

File: layouts/joomla/form/field/color/advanced.php (hunk 8 above)
Comment (thednp) [reply layouts/joomla/form/field/color/advanced.php]: Roger that, I hope that doesn't apply to JS as well.

**Comment by thednp** [comment]
@richard67 I cannot.

File: layouts/joomla/form/field/color/advanced.php (hunk 7 above)
Comment (richard67) [reply layouts/joomla/form/field/color/advanced.php]: No, Brian has only suggested, he hasn't changed anything.

Comment (thednp) [reply layouts/joomla/form/field/color/advanced.php]: @richard67 I don't actually know which one is correct.

Comment (thednp) [reply layouts/joomla/form/field/color/advanced.php]: Another files has ` * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>`

Comment (thednp) [reply layouts/joomla/form/field/color/advanced.php]: What about now?

Comment (richard67) [reply layouts/joomla/form/field/color/advanced.php]: You simply shouldn't have touched it. Anyway your branch needs to be updated, it makes no sense to fix that before. But the fact that you are using the 4.0-dev branch of your repo and not another one makes it a bit more complicated for me to do that for you.

**Comment by richard67** [comment]
@thednp Please stop making changes for a while, I soon update your branch to fix conflicts. Stay tuned and wait until I report back here that it's done.

**Comment by richard67** [comment]
@thednp A few things:

1. The branch which you have used to make this PR was stone age, I had to merge over 1500 commits into it.
2. We do not change the copyrights anymore in files every year. Nut with your last change you had changed it from 2016 to 2021. Since the file was created in 2016 and not 2021, I have reverted that.
3. In general you should make yourself familiar with our code style rules, and that does not need to read big documentation. It just needs to adjust your editor or IDE so you can see spaces and tabs, and to have a look on the existing code around the places which you want to change, and adapt to that. It makes no sense to do everything different to the existing code, even if you think it is better.
4. If you click the link "Details" at the right side of the row with the "continuous-integration/drone/pr" test, you come to a page where at the left hand side is a list of tests. One of them is called PHPCS. If that is red, you will see the details on PHP code style errors reported by the linter. Those tell you what needs to be fixed.

File: layouts/joomla/form/field/color/advanced.php [hunk 9]
```diff
@@ -42,56 +43,137 @@
...
+// both tinycolor and <color-picker> handle alpha automatically
+// an invalid format defaults to 'hex'
+$format = in_array($format, array('hex', 'rgb', 'hsl', 'rgba', 'hsla'), true) ? str_replace('a', '', $format) : 'hex';
+
+// set a generic placeholder
+if ($validate !== 'color' && $format === 'rgb')
+{
+	$placeholder = 'rgba(0, 0, 0, 0.5) / rgb(0, 0, 0)';
+}
+else if ($validate !== 'color' && $format === 'hsl')
```
Comment (richard67) [inline layouts/joomla/form/field/color/advanced.php]:
```suggestion
elseif ($validate !== 'color' && $format === 'hsl')
```

**Comment by richard67** [comment]
@dgrammatiko Please keep secret that I can fix code conflicts ... I don't wanna have too many requests 😉

**Comment by thednp** [comment]
Too late, I already know :)

**Comment by richard67** [comment]
@thednp Now drone fails with compiling the SCSS, see here: https://ci.joomla.org/joomla/joomla-cms/40329/1/7 . Maybe @dgrammatiko can advise?

```
Cleanup the Vendor 
Recreating the media folder...
Error: Cannot find module '@ctrl/tinycolor/package.json'
Require stack:
- /********/src/build/build-modules-js/init/localise-packages.es6.js
- /********/src/build/build.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.resolve (internal/modules/cjs/helpers.js:94:19)
    at resolvePackage (/********/src/build/build-modules-js/init/localise-packages.es6.js:42:34)
    at module.exports.localisePackages (/********/src/build/build-modules-js/init/localise-packages.es6.js:138:19)
    at async /********/src/build/build.js:163:7 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/********/src/build/build-modules-js/init/localise-packages.es6.js',
    '/********/src/build/build.js'
  ]
}
```

**Comment by thednp** [comment]
That is surely not my part. I mean I don't know how to fix it :)

**Comment by thednp** [comment]
For `tinycolor` I precompiled an es6 source from typescript and uploaded as instructed by @dgrammatiko. Other than that I don't know what's what. He said for tinycolor we need some new compile script but I don't actually know the status of that tbh.

Thanks again @richard67 have a good one, I hope you'll love this commit when done.

**Comment by dgrammatiko** [comment]
@thednp I think you created this PR solely on Github UI which will not work in this case (and fwiw most cases). You need to ingtall Git, then `git clone git@github.com:joomla/joomla-cms.git`.
Then you need to run composer and npm in that directory https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment

Then you create a new branch locally with all the changes, publish the branch and then you will be able to make a PR to the Joomla repo.

What is wrong here is that you said to the tools to copy a file from `node_modules` but you never installed `@ctrl/tinycolor` so the package.json and package.json.lock don;t have the dependency and thus everything fails, as expected. Follow the instructions above

**Comment by thednp** [comment]
I will do it again, differently.

**Comment by dgrammatiko** [comment]
> I will do it again, differently.

To save you from wasting your time (and ours) please:
- make sure that you install the dependency in your CLI (like `npm install -S @ctrl/tinycolor`
- follow my proposal https://github.com/thednp/joomla-color-picker/commit/13eec3f51a5ecb1158f3b43e93e221010e4156b9

**Comment by thednp** [comment]
@dgrammatiko the thing is: I've set the `v4-dev` branch as default, that was the problem in the first place, then when switching back, there goes another...

I'm going to save files I've worked on the branch in my local computer, delete this entire fork and start a new fresh.

**Comment by dgrammatiko** [comment]
> I'm going to save files I've worked on the branch in my local computer, delete this entire fork and start a new fresh.

I made a mistake on my previous comment, you shouldn't clone the Joomla repo but your fork `git clone git@github.com:thednp/joomla-color-picker.git`change to the branch to `4.0-dev`  then do the composer,npm stuff, etc

**Comment by thednp** [comment]
I don't know how to use the git app, I use TurtoiseSVN for my repos, and the browser for the other things.

**Pull request closed without merging**